# CMU-MOSI Multimodal SDK

# Download and Process mosi

### word level align

In [2]:
import mmsdk
from mmsdk import mmdatasdk
import numpy

In [3]:
#A simple averaging technique. More advanced methods can be built based on intervals.
def myavg(intervals,features):
        return numpy.average(features,axis=0)

In [ ]:
#Downloading the dataset
cmumosi_highlevel=mmdatasdk.mmdataset(mmdatasdk.cmu_mosi.highlevel,'cmumosi/')
# cmumosi_highlevel=mmdatasdk.mmdataset('cmumosi/')

In [6]:
#some random video from cmumosi_highlevel
some_video=list(cmumosi_highlevel["glove_vectors"].data.keys())[0]

In [ ]:
#The next line is not needed for standard datasets as they are all sorted based on intervals in computational sequence entries
cmumosi_highlevel.sort()

In [ ]:
#Aligning to the words to get word-level alignments
cmumosi_highlevel.align('glove_vectors',collapse_functions=[myavg])
cmumosi_highlevel.impute('glove_vectors')

In [9]:
#get the intervals and features accompanying the 100th word in the some_video
some_video_100th_word=some_video+'[100]'
for compseq_name in list(cmumosi_highlevel.computational_sequences.keys()):
	compseq=cmumosi_highlevel[compseq_name]
	print (compseq_name)
	print (numpy.array(compseq.data[some_video_100th_word]["intervals"]).shape,numpy.array(compseq.data[some_video_100th_word]["features"]).shape)
	print ("-------")

glove_vectors
(1, 2) (1, 300)
-------
FACET_4.1
(1, 2) (1, 47)
-------
FACET_4.2
(1, 2) (1, 35)
-------
OpenSmile-emobase2010
(1, 2) (1, 1585)
-------
OpenSMILE
(1, 2) (1, 384)
-------
OpenFace_1
(1, 2) (1, 430)
-------
OpenFace_2
(1, 2) (1, 713)
-------
COVAREP
(1, 2) (1, 74)
-------


k번째 비디오의 100번째 단어의 computational sequence를 관찰한 결과,

* Text - glove_vectors (feature: 300)

* Image - OpenFace_1 (430), FACET_4.2 (35), FACET_4.1 (47), COVAREP(74), OpenFace_2 (713)

* Audio - OpenSmile_emobase2010 (1585), b'OpenSMILE (384)

이를 통해 embedding feature 수 파악 가능.

*참고로 k번째 비디오는 glove_vector로 리스트업 하여 추출한 것.*

```python
#some random video from cmumosi_highlevel
some_video=list(cmumosi_highlevel["glove_vectors"].data.keys())[0]
```

![computational sequence](examples/compseq.png)

2번째 비디오를 관찰해보아도 마찬가지의 결과

In [10]:
#get the intervals and features accompanying the 2nd in some_video
some_video_2nd_segment=some_video+'[2]'
for compseq_name in list(cmumosi_highlevel.computational_sequences.keys()):
	compseq=cmumosi_highlevel[compseq_name]
	print (compseq_name)
	print (numpy.array(compseq.data[some_video_2nd_segment]["intervals"]).shape,numpy.array(compseq.data[some_video_2nd_segment]["features"]).shape)
	print ("-------")

glove_vectors
(1, 2) (1, 300)
-------
FACET_4.1
(1, 2) (1, 47)
-------
FACET_4.2
(1, 2) (1, 35)
-------
OpenSmile-emobase2010
(1, 2) (1, 1585)
-------
OpenSMILE
(1, 2) (1, 384)
-------
OpenFace_1
(1, 2) (1, 430)
-------
OpenFace_2
(1, 2) (1, 713)
-------
COVAREP
(1, 2) (1, 74)
-------


In [11]:
#Deploying the files to the disk and reading them again - Building machine learning models start right after this. No need to do alignment multiple times since aligned files can be deployed and used again.
# deploy_files={x:x for x in cmumosi_highlevel.computational_sequences.keys()}
# cmumosi_highlevel.deploy("./deployed",deploy_files)

# #Reading the dumped file can be as easy as just calling the mmdataset on the deployed folder
# aligned_cmumosi_highlevel=mmdatasdk.mmdataset('./deployed')

#Now let's get the tensor ready for ML - right here we just get everything into ML ready tensors. But you should split the aligned_cmumosi_highlevel based on the standard CMU MOSI folds
#get the standard folds using mmsdk.mmdatasdk.cmu_mosi.standard_folds.standard_x_fold for x={train,test,valid}


In [ ]:
import tensorflow as tf

Error: Session cannot generate requests

In [ ]:
tensors=cmumosi_highlevel.get_tensors(seq_len=25,non_sequences=["Opinion Segment Labels"],direction=False,folds=[mmdatasdk.cmu_mosi.standard_folds.standard_train_fold,mmdatasdk.cmu_mosi.standard_folds.standard_valid_fold,mmdatasdk.cmu_mosi.standard_folds.standard_test_fold])

fold_names=["train","valid","test"]

for i in range(3):
	for csd in list(cmumosi_highlevel.keys()):
		print ("Shape of the %s computational sequence for %s fold is %s"%(csd,fold_names[i],tensors[i][csd].shape))


Error: Session cannot generate requests